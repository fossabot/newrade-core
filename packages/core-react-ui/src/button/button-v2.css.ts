const iconSize = `1.5em`;
const iconPadding = `3em`;

/**
 * Base
 */

import { style } from '@vanilla-extract/css';

import { buttonsVars } from '../theme/default-theme.css';

export const base = style({
  display: 'block',
  position: 'relative',
  outline: 'none',
  lineHeight: 0,
  cursor: 'pointer',
  appearance: 'none',
  userSelect: 'none',
  touchAction: 'none',
  WebkitAppearance: 'none',
  whiteSpace: 'nowrap', // don't allow wrapping
  borderRadius: 4,
  width: 'fit-content',
  ':focus': {
    // boxShadow: `0 0 0 1px ${cssTheme.colors.colors.primary[700]}`,
  },
});

/**
 * States
 */
export const rest = style({});
export const loading = style({});
export const disabled = style({});

/**
 *
 * Variants
 *
 */

export const primary = style([
  base,
  style({
    color: buttonsVars.buttons.variants.primary.textColor,
    backgroundColor: buttonsVars.buttons.variants.primary.backgroundColor,
  }),
]);

/**
 *
 * Sizes
 *
 */

export const large = style({
  padding: buttonsVars.buttons.sizes.medium.padding.default,
  borderWidth: buttonsVars.buttons.sizes.large.border.default.width,
});

export const medium = style({
  padding: buttonsVars.buttons.sizes.medium.padding.default,
  borderWidth: buttonsVars.buttons.sizes.medium.border.default.width,
});

export const small = style({
  padding: buttonsVars.buttons.sizes.medium.padding.default,
});

export const xSmall = style({
  padding: buttonsVars.buttons.sizes.medium.padding.default,
});
