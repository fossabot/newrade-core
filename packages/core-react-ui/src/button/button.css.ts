import { createVar, style } from '@vanilla-extract/css';
import { recipe, RecipeVariants } from '@vanilla-extract/recipes';

import { colorVars, layoutCSS, typographyVars } from '../default-theme/default-theme.css';

/**
 *
 * Vars
 *
 */

const iconSize = createVar();
const iconMargin = createVar();
const iconColor = createVar();
const iconColorDisabled = createVar();

const paddingX = createVar();
const paddingY = createVar();
const height = createVar();

const borderWidth = createVar();
const borderColor = createVar();
const borderColorActive = createVar();
const borderColorDisabled = createVar();
const borderRadius = createVar();

const backgroundColor = createVar();
const backgroundColorActive = createVar();
const backgroundColorDisabled = createVar();

const textColor = createVar();
const textColorActive = createVar();
const textColorDisabled = createVar();

/**
 *
 * Variants
 *
 */

export const variants = recipe({
  base: {
    /**
     *
     * Reset and base CSS
     *
     */

    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    position: 'relative',
    outline: 'none',
    lineHeight: 0,
    cursor: 'pointer',
    appearance: 'none',
    userSelect: 'none',
    touchAction: 'none',
    WebkitAppearance: 'none',
    textDecoration: 'none',

    /**
     *
     * Box model
     *
     */

    width: 'fit-content',
    minWidth: height,
    height: height,
    paddingLeft: paddingX,
    paddingRight: paddingX,
    margin: 0,
    borderStyle: borderStyle,
    borderWidth: borderWidth,
    borderRadius: borderRadius,
    outline: 'none',

    /**
     *
     * Content
     *
     */

    // don't text or content to wrap
    whiteSpace: 'nowrap',

    /**
     *
     * Colors
     *
     */

    color: textColor,
    backgroundColor: backgroundColor,
    borderColor: borderColor,

    /**
     *
     * Effects
     *
     */

    /**
     *
     * Interaction & states
     *
     */

    cursor: 'pointer',
    transition: `transform 0.1s ease-in`,
    userSelect: 'none',
    touchAction: `pan-y pan-x`,

    selectors: {
      '&:active': {
        color: textColorActive,
        backgroundColor: backgroundColorActive,
        borderColor: borderColorActive,

        transform: `translateY(1px)`,
      },

      '&[disabled]': {
        color: textColorDisabled,
        backgroundColor: backgroundColorDisabled,
        borderColor: borderColorDisabled,

        cursor: 'not-allowed',
      },
    },
  },
  variants: {
    /**
     *
     * Variants
     *
     */

    variant: {
      primary: {
        vars: {
          [iconColor]: colorVars.colors.grey[700],
          [iconColorDisabled]: colorVars.colorIntents.disabledText,
          [textColor]: colorVars.colorIntents.primaryText,
          [textColorActive]: colorVars.colorIntents.primary,
          [textColorDisabled]: colorVars.colors.grey[600],
          [borderColor]: colorVars.colorIntents.primary,
          [borderColorDisabled]: `transparent`,
          [backgroundColor]: colorVars.colors.grey[0],
          [backgroundColorActive]: colorVars.colors.primary[100],
          [backgroundColorDisabled]: colorVars.colors.grey[25],
        },
      },
      secondary: {},
      tertiary: {},
    },
    /**
     *
     * Sizes
     *
     */

    size: {
      small: {
        vars: {
          [paddingX]: buttonsVars.sizes.large.padding.left,
          [height]: buttonsVars.sizes.large.height,
          [iconSize]: buttonsVars.sizes.large.iconSize,
          [iconMargin]: buttonsVars.sizes.large.iconMargin.right,
          [borderWidth]: buttonsVars.sizes.large.border.default.width,
          [borderStyle]: buttonsVars.sizes.large.border.default.style,
          [borderRadius]: buttonsVars.sizes.large.border.default.radius,
        },
        fontSize: typographyVars.labels.mobile.small.capsize.fontSize,
        '@media': {
          [layoutCSS.media.tablet]: {
            fontSize: typographyVars.labels.tablet.small.capsize.fontSize,
          },
          [layoutCSS.media.desktopSmall]: {
            fontSize: typographyVars.labels.desktop.small.capsize.fontSize,
          },
        },
      },
      medium: {
        vars: {
          [paddingX]: buttonsVars.sizes.medium.padding.left,
          [height]: buttonsVars.sizes.medium.height,
          [iconSize]: buttonsVars.sizes.medium.iconSize,
          [iconMargin]: buttonsVars.sizes.medium.iconMargin.right,
          [borderWidth]: buttonsVars.sizes.medium.border.default.width,
          [borderStyle]: buttonsVars.sizes.medium.border.default.style,
          [borderRadius]: buttonsVars.sizes.medium.border.default.radius,
        },
        fontSize: typographyVars.labels.mobile.small.capsize.fontSize,
        '@media': {
          [layoutCSS.media.tablet]: {
            fontSize: typographyVars.labels.tablet.small.capsize.fontSize,
          },
          [layoutCSS.media.desktopSmall]: {
            fontSize: typographyVars.labels.desktop.small.capsize.fontSize,
          },
        },
      },
      small: {
        vars: {
          [paddingX]: buttonsVars.sizes.small.padding.left,
          [height]: buttonsVars.sizes.small.height,
          [iconSize]: buttonsVars.sizes.small.iconSize,
          [iconMargin]: buttonsVars.sizes.small.iconMargin.right,
          [borderWidth]: buttonsVars.sizes.small.border.default.width,
          [borderStyle]: buttonsVars.sizes.small.border.default.style,
          [borderRadius]: buttonsVars.sizes.small.border.default.radius,
        },
        fontSize: typographyVars.labels.mobile.small.capsize.fontSize,
        '@media': {
          [layoutCSS.media.tablet]: {
            fontSize: typographyVars.labels.tablet.small.capsize.fontSize,
          },
          [layoutCSS.media.desktopSmall]: {
            fontSize: typographyVars.labels.desktop.small.capsize.fontSize,
          },
        },
      },
      xSmall: {
        vars: {
          [paddingX]: buttonsVars.sizes.xSmall.padding.left,
          [height]: buttonsVars.sizes.xSmall.height,
          [iconSize]: buttonsVars.sizes.xSmall.iconSize,
          [iconMargin]: buttonsVars.sizes.xSmall.iconMargin.right,
          [borderWidth]: buttonsVars.sizes.xSmall.border.default.width,
          [borderStyle]: buttonsVars.sizes.xSmall.border.default.style,
          [borderRadius]: buttonsVars.sizes.xSmall.border.default.radius,
        },
        fontSize: typographyVars.labels.mobile.small.capsize.fontSize,
        '@media': {
          [layoutCSS.media.tablet]: {
            fontSize: typographyVars.labels.tablet.small.capsize.fontSize,
          },
          [layoutCSS.media.desktopSmall]: {
            fontSize: typographyVars.labels.desktop.small.capsize.fontSize,
          },
        },
      },
    },
  },
});

/**
 *
 * Typography
 *
 */

export const text = style({
  fontFamily: typographyVars.labels.font,
  userSelect: 'none',
});

/**
 *
 * Base
 *
 */

export const styleVariants = recipe({
  base: {
    position: 'relative',
    padding: `0 calc(${iconSize} + 2 * ${iconOffset}) 0 ${paddingX}`,
    height: height,
    lineHeight: `calc(${height} - 2 * ${borderWidth})`,
    minWidth: `100%`,
    width: 'fit-content',
    verticalAlign: 'middle',
    ':focus': {
      boxShadow: `0 0 1px 1px ${colorVars.colors.primary[300]}`,
    },
    '::-webkit-input-placeholder': {
      color: colorVars.colors.grey[200],
      verticalAlign: 'middle',
      lineHeight: height,
    },

    borderStyle: 'solid',
    borderWidth: borderWidth,
    borderRadius: borderRadius,

    color: textColor,
    backgroundColor: backgroundColor,

    outline: 'none',
    whiteSpace: 'nowrap',

    cursor: 'auto',
    appearance: 'none',

    userSelect: 'auto',
    touchAction: 'none',
    WebkitAppearance: 'none',

    transition: `box-shadow 0.3s ease-out`,
  },
  variants: {
    /**
     * Variants
     */
    variant: {
      primary: {},
      secondary: {},
      tertiary: {},
    },
    /**
     * Sizes
     */
    size: {
      small: {},
      medium: {},
      large: {},
    },
    /**
     * Icons
     */
    icon: {
      left: { paddingLeft: `calc(2 * ${paddingX} + ${iconSize})` },
      right: { paddingRight: `calc(2 * ${paddingX} + ${iconSize})` },
    },
    /**
     * States (internal)
     */
    state: {
      rest: {},
      focused: {
        color: textColorActive,
        borderColor: borderColorActive,
        backgroundColor: backgroundColorActive,
      },
    },
    /**
     * Validity states
     */
    validity: {
      neutral: {},
      valid: {
        boxShadow: `0 0 0 1px ${colorVars.colorIntents.successAction}`,
      },
      warning: {
        boxShadow: `0 0 0 1px ${colorVars.colorIntents.warningAction}`,
      },
      error: {
        boxShadow: `0 0 0 1px ${colorVars.colorIntents.dangerAction}`,
      },
    },
    /**
     * Loading states
     */
    loading: {
      true: {},
      false: {},
    },
    /**
     * Disabled states
     */
    disabled: {
      true: {
        cursor: 'not-allowed',
        color: textColorDisabled,
        borderColor: borderColorDisabled,
        backgroundColor: backgroundColorDisabled,
        ':focus': {
          outline: 'none',
          boxShadow: 'none',
        },
      },
      false: {},
    },
  },
});

export type InputVariants = RecipeVariants<typeof styleVariants>;

/**
 *
 * Icons
 *
 */

export const iconVariants = recipe({
  base: {
    color: iconColor,
    position: 'relative',
    height: iconSize,
    width: iconSize,
    pointerEvents: 'none', // let events go through
  },
  variants: {
    /**
     * Icon position
     */
    position: {
      left: { marginRight: iconMargin },
      right: { marginLeft: iconMargin },
    },
    /**
     * Disabled states
     */
    disabled: {
      true: {
        color: iconColorDisabled,
      },
      false: {},
    },
  },
});

export type InputIconVariants = RecipeVariants<typeof iconVariants>;
