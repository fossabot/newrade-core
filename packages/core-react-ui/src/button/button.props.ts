import { AnchorHTMLAttributes } from 'react';

import { ButtonProps } from '@newrade/core-design-system';

import { PrimitiveProps } from '../primitive/primitive.props';

export type ButtonAsType = 'button' | 'a' | 'div';

export type ButtonCompProps = PrimitiveProps<ButtonAsType> &
  Pick<AnchorHTMLAttributes<any>, 'href' | 'target' | 'type'> &
  ButtonProps & {
    /**
     * Rendering element types
     */
    as?: ButtonAsType;
    /**
     * Pass custom svg icon
     */
    IconSVG?: React.ReactNode;
    /**
     * Remove padding on either side which is usefull to align a button's text or icon to a layout.
     */
    collapsePadding?: 'left' | 'right';
  };
