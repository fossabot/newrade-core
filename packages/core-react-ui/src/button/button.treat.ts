import { globalStyle, style } from 'treat';

import { Theme } from '../design-system';

const iconSize = `1.5em`;
const iconOnlySize = `1.8em`;
const iconPadding = `3em`;

/**
 * Base
 */
export const base = style(({ theme, cssTheme }: Theme) => ({
  display: 'block',
  position: 'relative',
  outline: 'none',
  lineHeight: 0,
  cursor: 'pointer',
  appearance: 'none',
  userSelect: 'none',
  touchAction: 'none',
  WebkitAppearance: 'none',
  whiteSpace: 'nowrap', // don't allow wrapping
  borderRadius: 4,
  width: 'fit-content',
  ':focus': {
    // boxShadow: `0 0 0 1px ${cssTheme.colors.colors.primary[700]}`,
  },
}));

/**
 * States
 */
export const rest = style(({ theme, cssTheme }: Theme) => ({}));
export const loading = style(({ theme, cssTheme }: Theme) => ({}));
export const disabled = style(({ theme, cssTheme }: Theme) => ({}));

/**
 *
 * Variants
 *
 */

/**
 * Primary
 */
export const primary = style(({ theme, cssTheme }: Theme) => ({
  color: cssTheme.components.buttons.var.variants.primary.textColor,
  border: `2px solid ${cssTheme.components.buttons.var.variants.primary.borderColor}`,
  backgroundColor: cssTheme.colors.var.colors.primary[500],
  selectors: {
    '&:active': {
      filter: `brightness(90%)`,
    },
    '&[disabled]': {
      cursor: 'not-allowed',
      color: cssTheme.colors.colorIntents.disabledText,
      borderColor: cssTheme.colors.colorIntents.backgroundDisabled,
      backgroundColor: cssTheme.colors.colorIntents.backgroundDisabled,
    },
  },
}));

/**
 * Secondary
 */
export const secondary = style(({ theme, cssTheme }: Theme) => ({
  color: cssTheme.colors.colorIntents.primary,
  border: `2px solid ${cssTheme.colors.colorIntents.primary}`,
  selectors: {
    '&:active': {
      backgroundColor: cssTheme.colors.colors.primary[100],
    },
    '&[disabled]': {
      cursor: 'not-allowed',
      color: cssTheme.colors.colorIntents.disabledText,
      borderColor: cssTheme.colors.colorIntents.backgroundDisabled,
      backgroundColor: cssTheme.colors.colorIntents.backgroundDisabled,
    },
  },
}));

/**
 * Tertiary
 */
export const tertiary = style(({ theme, cssTheme }: Theme) => ({
  color: cssTheme.colors.colorIntents.primary,
  selectors: {
    '&:active': {
      backgroundColor: cssTheme.colors.colors.primary[100],
    },
    '&[disabled]': {
      cursor: 'not-allowed',
      color: cssTheme.colors.colorIntents.disabledText,
      backgroundColor: cssTheme.colors.colorIntents.backgroundDisabled,
    },
  },
}));

/**
 * Sizes
 */
export const large = style(({ theme, cssTheme }: Theme) => ({
  height: `calc(2 * ${cssTheme.sizing.var.x3} + ${
    cssTheme.typography.labels.mobile.medium.capHeight
  }px + ${4}px)`,
  minWidth: `calc(2 * ${cssTheme.sizing.var.x3} + ${
    cssTheme.typography.labels.mobile.medium.capHeight
  }px + ${4}px)`,
  padding: `${cssTheme.sizing.var.x3} ${cssTheme.sizing.var.x3}`,
  fontSize: cssTheme.typography.labels.mobile.medium.capsize?.fontSize,
  borderRadius: cssTheme.components.buttons.var.sizes.large.border?.default?.radius,
  '@media': {
    [cssTheme.layout.media.tablet]: {
      height: `calc(2 * ${cssTheme.sizing.var.x3} + ${
        cssTheme.typography.labels.tablet.medium.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x3} + ${
        cssTheme.typography.labels.tablet.medium.capHeight
      }px + ${4}px)`,
      fontSize: cssTheme.typography.labels.tablet.medium.capsize?.fontSize,
    },
    [cssTheme.layout.media.desktopSmall]: {
      height: `calc(2 * ${cssTheme.sizing.var.x3} + ${
        cssTheme.typography.labels.desktop.medium.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x3} + ${
        cssTheme.typography.labels.desktop.medium.capHeight
      }px + ${4}px)`,
      fontSize: cssTheme.typography.labels.desktop.medium.capsize?.fontSize,
    },
  },
  selectors: {
    '&[datapaddingcollapse="left"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x3} + (${iconSize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[datapaddingcollapse="left-icon"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x3} + (${iconOnlySize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[dataicon="right"]': {
      padding: `${cssTheme.sizing.var.x3} ${iconPadding} ${cssTheme.sizing.var.x3} ${cssTheme.sizing.var.x3}`,
    },
    '&[dataicon="left"]': {
      padding: `${cssTheme.sizing.var.x3} ${cssTheme.sizing.var.x3} ${cssTheme.sizing.var.x3} ${iconPadding}`,
    },
  },
}));
export const collapsePaddingLeft = style(({ theme, cssTheme }: Theme) => ({}));
export const medium = style(({ theme, cssTheme }: Theme) => ({
  height: `calc(2 * ${cssTheme.sizing.var.x2} + ${
    cssTheme.typography.labels.mobile.small.capHeight
  }px + ${4}px)`,
  minWidth: `calc(2 * ${cssTheme.sizing.var.x2} + ${
    cssTheme.typography.labels.mobile.small.capHeight
  }px + ${4}px)`,
  padding: `${cssTheme.sizing.var.x2} ${cssTheme.sizing.var.x2}`,
  fontSize: cssTheme.typography.labels.mobile.small.capsize?.fontSize,
  '@media': {
    [cssTheme.layout.media.tablet]: {
      height: `calc(2 * ${cssTheme.sizing.var.x2} + ${
        cssTheme.typography.labels.tablet.small.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x2} + ${
        cssTheme.typography.labels.tablet.small.capHeight
      }px + ${4}px)`,
      fontSize: cssTheme.typography.labels.tablet.small.capsize?.fontSize,
    },
    [cssTheme.layout.media.desktopSmall]: {
      height: `calc(2 * ${cssTheme.sizing.var.x2} + ${
        cssTheme.typography.labels.desktop.small.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x2} + ${
        cssTheme.typography.labels.desktop.small.capHeight
      }px + ${4}px)`,
      fontSize: cssTheme.typography.labels.desktop.small.capsize?.fontSize,
    },
  },
  selectors: {
    '&[datapaddingcollapse="left"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x2} + (${iconSize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[datapaddingcollapse="left-icon"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x2} + (${iconOnlySize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[dataicon="right"]': {
      padding: `${cssTheme.sizing.var.x2} ${iconPadding} ${cssTheme.sizing.var.x2} ${cssTheme.sizing.var.x2} `,
    },
    '&[dataicon="left"]': {
      padding: `${cssTheme.sizing.var.x2} ${cssTheme.sizing.var.x2} ${cssTheme.sizing.var.x2} ${iconPadding}`,
    },
  },
}));
export const small = style(({ theme, cssTheme }: Theme) => ({
  height: `calc(2 * ${cssTheme.sizing.var.x1} + ${
    cssTheme.typography.labels.mobile.small.capHeight
  }px + ${4}px)`,
  minWidth: `calc(2 * ${cssTheme.sizing.var.x1} + ${
    cssTheme.typography.labels.mobile.small.capHeight
  }px + ${4}px)`,
  padding: `${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x2}`,
  fontSize: cssTheme.typography.labels.mobile.small.capsize?.fontSize,
  '@media': {
    [cssTheme.layout.media.tablet]: {
      height: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.tablet.small.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.tablet.small.capHeight
      }px + ${4}px)`,
      fontSize: cssTheme.typography.labels.tablet.small.capsize?.fontSize,
    },
    [cssTheme.layout.media.desktopSmall]: {
      height: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.desktop.small.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.desktop.small.capHeight
      }px + ${4}px)`,
      fontSize: cssTheme.typography.labels.desktop.small.capsize?.fontSize,
    },
  },
  selectors: {
    '&[datapaddingcollapse="left"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x2} + (${iconSize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[datapaddingcollapse="left-icon"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x2} + (${iconOnlySize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[dataicon="right"]': {
      padding: `${cssTheme.sizing.var.x1} ${iconPadding} ${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x2}`,
    },
    '&[dataicon="left"]': {
      padding: `${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x2} ${iconPadding}`,
    },
  },
}));
export const xSmall = style(({ theme, cssTheme }: Theme) => ({
  height: `calc(2 * ${cssTheme.sizing.var.x1} + ${
    cssTheme.typography.labels.mobile.xSmall.capHeight
  }px + ${4}px)`,
  minWidth: `calc(2 * ${cssTheme.sizing.var.x1} + ${
    cssTheme.typography.labels.mobile.xSmall.capHeight
  }px + ${4}px)`,
  padding: `${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x1}`,
  fontSize: cssTheme.typography.labels.mobile.xSmall.capsize?.fontSize,
  '@media': {
    [cssTheme.layout.media.tablet]: {
      fontSize: cssTheme.typography.labels.tablet.xSmall.capsize?.fontSize,
    },
    [cssTheme.layout.media.desktopSmall]: {
      fontSize: cssTheme.typography.labels.desktop.xSmall.capsize?.fontSize,
    },
  },
  selectors: {
    '&[datapaddingcollapse="left"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x1} + (${iconSize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[datapaddingcollapse="left-icon"]': {
      marginLeft: `calc(-1 * ${cssTheme.sizing.var.x1} + (${iconOnlySize} - ${cssTheme.sizing.var.x3}) / 2)`,
    },
    '&[dataicon="right"]': {
      height: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.tablet.xSmall.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.tablet.xSmall.capHeight
      }px + ${4}px)`,
      padding: `${cssTheme.sizing.var.x1} ${iconPadding} ${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x1}`,
    },
    '&[dataicon="left"]': {
      height: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.desktop.xSmall.capHeight
      }px + ${4}px)`,
      minWidth: `calc(2 * ${cssTheme.sizing.var.x1} + ${
        cssTheme.typography.labels.desktop.xSmall.capHeight
      }px + ${4}px)`,
      padding: `${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x1} ${cssTheme.sizing.var.x1} ${iconPadding}`,
    },
  },
}));

/**
 * Icon
 */
export const iconBase = style(({ theme, cssTheme }: Theme) => ({
  position: 'absolute',
  height: iconSize,
  maxHeight: iconSize,
  width: iconSize,
  maxWidth: iconSize,
}));
export const iconOnly = style(({ theme, cssTheme }: Theme) => ({
  position: 'absolute',
  height: iconOnlySize,
  maxHeight: iconOnlySize,
  width: iconOnlySize,
  maxWidth: iconOnlySize,
}));
export const icon = style(({ theme, cssTheme }: Theme) => ({
  top: `calc(50% - ${iconOnlySize} / 2)`,
  left: `calc(50% - ${iconOnlySize} / 2)`,
}));
export const right = style(({ theme, cssTheme }: Theme) => ({
  top: `calc(50% - ${iconSize} / 2)`,
  right: `0.8em`,
}));
export const left = style(({ theme, cssTheme }: Theme) => ({
  top: `calc(50% - ${iconSize} / 2)`,
  left: `0.8em`,
}));
