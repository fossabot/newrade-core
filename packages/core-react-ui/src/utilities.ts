export * from './utilities/colors.utilities';
export * from './utilities/component.utilities';
export * from './utilities/css-variable.utilities';
export * from './utilities/effects.utilities';
export * from './utilities/filter-color.utility';
export * from './utilities/log.utilities';
export * from './utilities/text-capsize.utilities';
export * from './utilities/text.utilities';
export * from './utilities/theme.utilities';
export * from './utilities/utilities';

