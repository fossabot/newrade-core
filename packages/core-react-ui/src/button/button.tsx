import React, { useRef } from 'react';

import { ButtonIcon, ButtonSize, TEXT_STYLE, Variant } from '@newrade/core-design-system';

import { useCommonProps } from '../hooks/use-common-props.hook';
import { usePreventPinchZoom } from '../hooks/use-prevent-pinch-zoom';
import { Label } from '../text/label';
import { getDefaultTextFromProps, getMergedClassname } from '../utilities-iso';

import { ButtonCompProps } from './button.props';
import { getLabelSizeForButtonSize } from './button.utilities';

import * as styles from './button.css';

type Props = ButtonCompProps;

const defaultProps: Props = {
  size: ButtonSize.medium,
};

export const Button = React.forwardRef<any, Props>(function Button(
  {
    id,
    style,
    className,
    role,
    children,
    variant,
    collapsePadding,
    as,
    AsElement,
    size = defaultProps.size,
    disabled,
    loading,
    icon,
    Icon,
    IconSVG,
    ...props
  },
  forwardedRef
) {
  /**
   *
   * Ref forwarding
   *
   */

  const localRef = useRef<HTMLButtonElement>(null);
  const ref = forwardedRef ? (forwardedRef as React.RefObject<HTMLButtonElement>) : localRef;

  /**
   *
   * Event handling
   *
   */

  usePreventPinchZoom(ref.current);
  // usePreventLongPress(ref.current);

  /**
   *
   * Icon
   *
   */

  const dataicon = Icon ? (icon ? icon : ButtonIcon.right) : ButtonIcon.none;

  const iconClassNames = getMergedClassname([
    dataicon === ButtonIcon.icon ? styles.iconOnly : styles.iconBase,
    dataicon === ButtonIcon.icon ? styles.icon : '',
    dataicon === ButtonIcon.right ? styles.right : '',
    dataicon === ButtonIcon.left ? styles.left : '',
  ]);

  const IconSvg = IconSVG
    ? React.cloneElement(IconSVG as React.ReactElement, {
        className: iconClassNames,
        preserveAspectRatio: `xMinYMin meet`,
      })
    : null;

  /**
   *
   * Rendering
   *
   */

  const type = as ? as : 'button';

  const variantStateClassName = `${styles.base}`;
  const variantClassName = `${styles[variant ? variant : Variant.primary]}`;
  const variantSizeClassName = styles[size ? size : ButtonSize.medium];
  const collapsePaddingProp =
    dataicon === ButtonIcon.icon ? `${collapsePadding}-icon` : collapsePadding;

  const commonProps = useCommonProps<'button'>({
    id,
    style,
    className,
    role: as === 'button' ? '' : role,
    classNames: [variantStateClassName, variantSizeClassName, variantClassName, className],
    // @ts-ignore
    dataicon: dataicon,
    datapaddingcollapse: collapsePaddingProp,
    ...props,
  });

  const renderedChildren = children
    ? children
    : dataicon === ButtonIcon.icon
    ? ' '
    : getDefaultTextFromProps('button', {
        variant,
        size,
        icon,
        disabled,
      });

  const CustomElement = AsElement
    ? React.cloneElement(
        AsElement as React.ReactElement,
        commonProps,
        <>
          {icon === ButtonIcon.icon ? null : (
            <Label
              variantDisplay={'inline'}
              textStyle={TEXT_STYLE.bold}
              variant={getLabelSizeForButtonSize(size)}
            >
              {renderedChildren}
            </Label>
          )}
          {IconSvg}
        </>
      )
    : null;

  if (CustomElement) {
    return CustomElement;
  }

  const CustomElementAs =
    type !== 'button'
      ? React.createElement(
          type,
          commonProps,
          <>
            {icon === ButtonIcon.icon ? null : (
              <Label
                variantDisplay={'inline'}
                textStyle={TEXT_STYLE.bold}
                variant={getLabelSizeForButtonSize(size)}
              >
                {renderedChildren}
              </Label>
            )}
            {IconSvg}
          </>
        )
      : null;

  if (CustomElementAs) {
    return CustomElementAs;
  }

  return (
    <button ref={ref} disabled={disabled} {...commonProps}>
      {icon === ButtonIcon.icon ? null : (
        <Label
          variantDisplay={'inline'}
          textStyle={TEXT_STYLE.bold}
          variant={getLabelSizeForButtonSize(size)}
        >
          {renderedChildren}
        </Label>
      )}
      {IconSvg}
    </button>
  );
});
