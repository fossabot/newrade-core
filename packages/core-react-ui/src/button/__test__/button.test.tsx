import React from 'react';

import { renderTest } from '../../utilities-test/test-render.utilities';
import { Button } from '../button';

describe('the button component', () => {
  it('should render correctly', () => {
    const { container } = renderTest(<Button>Hello world</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
