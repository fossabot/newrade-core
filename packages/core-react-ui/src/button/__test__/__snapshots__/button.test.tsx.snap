// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the button component should render correctly 1`] = `
<div>
  <button
    class="button__x17q3ri"
    dataicon="none"
  >
    <div
      class="label__1i57ob90  text-color_variant_primary__11gow270  label_size_small__1i57ob93 label_style_bold__1i57ob96"
    >
      Hello world
    </div>
  </button>
</div>
`;
